src/graphapi/opgraph.cpp

```diff
@@ -0,0 +1,107 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/graphapi/opgraph.h>
+
+namespace miopen {
+namespace graphapi {
+
+OpNode::~OpNode() = default;
+
+OpGraph OpGraphBuilder::build()
+{
+
+    OpGraph graph;
+
+    // key = tensor ptr, value = vec. of dest nodes
+    std::unordered_map<Tensor*, EdgeInfo> e_map;
+
+    for(OpNode* n : mNodes)
+    {
+
+        for(Tensor* i : n->getInTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.emplace(i, EdgeInfo{}); // add empty EdgeInfo if not present
+
+            iter->second.mDests.emplace_back(n);
+        }
+
+        for(Tensor* o : n->getOutTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.emplace(o, EdgeInfo{}); // add empty EdgeInfo if not present
+
+            assert(iter->second.mSrc == nullptr);
+            iter->second.mSrc = n;
+        }
+    }
+
+    graph.addNodes(std::move(mNodes));
+
+    for(const auto& [tens_ptr, edge_info] : e_map)
+    {
+
+        if(edge_info.mSrc != nullptr && !edge_info.mDests.empty())
+        {
+            for(const auto& d : edge_info.mDests)
+            {
+                graph.addEdge(edge_info.mSrc, tens_ptr, d);
+            }
+        }
+        else if(edge_info.mSrc == nullptr)
+        {
+
+            // tens_ptr is a non-virtual input tensor
+            assert(!tens_ptr->isVirtual()); // tensor pointer can't be virtual
+            assert(!edge_info.mDests.empty());
+
+            // NOTE(Amber): we may take out this step if we decide not to add a source
+            // and sink in the graph
+            for(const auto& d : edge_info.mDests)
+            {
+                graph.addEdgeFromSrc(d, tens_ptr);
+            }
+        }
+        else if(edge_info.mDests.empty())
+        {
+            // tens_ptr is a non-virtual output tensor
+            assert(!tens_ptr->isVirtual());
```

Comment:
This one is critical. The other comments are less critical and I don't care much about them.

It can (and will) be virtual.
Forward fp8-mha uses virtual output nodes to reflect that it works in inference mode.

Moreover we must be able to match one algorithm to the both cases. Probably it can be worked around by advertising two different signatures.

Reply:
Yeah, I did realize after you mentioned it that an output tensor can be virtual. I will remove the `assert` to allow it. 
In general, `virtual/non-virtual` will not be part of the graph signature (as of now at least).